---
title: "Coverage Analysis: Entamoeba histolytica knockout libraries"
author: "Sam Hunter"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
    html_document:
        keep_md: yes
fig_width: 8
fig_height: 10

---

```{r global_options, include=FALSE}
knitr::opts_chunk$set(fig.width=14, fig.height=12, fig.path='Figs/',
                      echo=FALSE, warning=FALSE, message=FALSE)
```

```{r setup, include=FALSE}
#install.packages("data.table")
#install.packages("tidyverse")

# library(data.table)
# library(tidyverse)
# library(pheatmap)
library(kableExtra)
library(Biostrings)
library(Rsubread)
#library(cowplot)
library(readr)
library(ggplot2)
library(pheatmap)


options(stringsAsFactors=F)

nice_colors = c("#999999", "#E69F00", "#56B4E9","#e98756","#c08160","#5800e6", "#CDDC49",
                "#C475D3", "#E94B30", "#233F57", "#FEE659", "#A1CFDD", "#F4755E", "#D6F6F7","#EB6D58", "#6898BF")

```

```{r functions, echo=FALSE}

plotSummaries = function(fcount, count_method, mdata){
    df = data.frame(Sample = gsub('.bam', '', colnames(fcount$stat)[2:ncol(fcount$stat)]),
        AssignedReads = unlist(fcount$stat[fcount$stat$Status == 'Assigned', 2:ncol(fcount$stat), ]),
        CoveredGenes = colSums(fcount$counts>0)
                )
    df = data.frame(df, mdata[match(df$Sample, mdata$Sample), ])                
    df = data.frame(df, readcount[match(df$Sample, readcount$Sample), ])
    df = df[df$Sample != 'Undetermined', ]

    # TODO
    cat("### Summary statistics for all samples:\n")
    cat("<br>")
    print(kable(df[,c("Sample","reads", "AssignedReads","CoveredGenes","Fragmented.gDNA", "Group", "Type", "Rep", "PCR.Cycles")], padding=2) %>%
    kable_styling("striped", full_width = F) %>%
    row_spec(0, angle = 0))

    # # Combine replicates and present summary stats:
    # cat("<br>")
    # cat("<br>")
    # cat("### Summary statistics for combined samples:\n")
    # cat("<br>")
    # df = 

    cat("<br>")
    cat("<br>")
    cat("### How does gene coverage correspond to sequencing depth?\n")
    cat("<br>")

    gp1 = ggplot(df, aes(y=CoveredGenes, x=reads, color=Group, shape=Type, label=Sample) ) + 
        geom_point(size=2) + 
        scale_color_manual(values=nice_colors[-1]) + theme_bw() +
        geom_hline(yintercept = 8333, size=1, color = 'tan2')+
        ggtitle(paste("Coverage in", count_method, "read orientation")) +
        #theme(plot.title = element_text(hjust = 0.5)) +
        xlab('Reads Sequenced') + 
        ylab('Genes covered by at least one read with \u2265 27bp overlap')

    print(gp1)

    # What sort of per-gene coverage is there?
    cat("<br>")
    cat("<br>")
    cat("<br>")
    cat("What is the distribution of number of reads mapped per gene?\n")
    cat("<br>")

    counts_stacked = data.frame(stack(fcount$counts[,colnames(fcount$counts) != 'Undetermined.bam']))
    counts_stacked$col = gsub('.bam', '', counts_stacked$col)
    counts_stacked$Group = mdata$Group[match(counts_stacked$col, mdata$Sample)]
    counts_stacked$Type = mdata$Type[match(counts_stacked$col, mdata$Sample)]

    gp2 = ggplot(counts_stacked) + 
    geom_boxplot(aes(x=col, y=value, fill=Type))+
    scale_fill_manual(values=nice_colors[-1]) + theme_bw() +
    scale_y_continuous(limits = c(0, 100)) +
    theme(axis.text.x = element_text(angle = 90)) +
    theme(plot.title = element_text(hjust = 0.5)) +
    ggtitle(paste("Reads mapped in", count_method, "orientation per gene")) +
    xlab('') + 
    ylab('Read Count, \u2265 27bp overlap') + facet_wrap(. ~ Group, scales="free_x")

    suppressWarnings(print(gp2))

    # Read coverage across all genes, heatmap:
    cat("<br>")
    cat("<br>")
    cat("What is read coverage like across all genes, and do samples cluster by prep?\n")
    cat("<br>")
    
    gcounts = fcount$counts[,colnames(fcount$counts) != 'Undetermined.bam']
    colnames(gcounts) = gsub('.bam', '', colnames(gcounts))
    anncol = mdata[,c('Group', 'Type')]
    rownames(anncol) = mdata$Sample
    gp3 = pheatmap(log2(gcounts+1), main='Log2 read coverage by gene', show_rownames=F, ylab='Genes', 
        cluster_rows=F, scale='none',annotation_col=anncol, fontsize=7, cutree_cols = ncol(gcounts))
    

    # Read coverage across all genes for Final samples only.
    cat("<br>")
    cat("<br>")
    cat("What is read coverage like across all genes for Final samples only? Do samples cluster by prep?\n")
    cat("<br>")

    finalSamples = mdata$Sample[as.character(mdata$Group) == 'Final']
    gcountsF = fcount$counts[,paste0(finalSamples, '.bam')]
    colnames(gcountsF) = gsub('.bam', '', colnames(gcountsF))
    anncolF = mdata[match(finalSamples, mdata$Sample) ,c('Group', 'Type')]
    rownames(anncolF) = finalSamples
    gp4 = pheatmap(log2(gcountsF+1), main='Final Samples: Log2 read coverage by gene', show_rownames=F, ylab='Genes', 
        cluster_rows=F, scale='none',annotation_col=anncolF, fontsize=7, cutree_cols = ncol(gcountsF))

    # cat("<br>")
    # cat("<br>")
    # cat("How consistently are genes targeted across samples?\n")
    # cat("<br>")
    # df = rowSums(gcounts>0) 
    # barplot(table(rowSums(gcounts > 0)), main="Consistency of gene coverage", 
    #     ylab='Number of Genes', xlab="Number of samples genes are targeted in")


    # TODO convert barplot to ggplot
    # ggplot(ffps, aes(fill=Orientation, y=value, x=sample)) +
    #     facet_wrap(. ~ Group, drop=TRUE, scales='free') +
    #     geom_bar(stat="identity", alpha=.7, color='black') + 
    #     scale_fill_manual(values=nice_colors) + theme_bw() +
    #     theme(axis.text.x = element_text(angle = 90)) +
    #     xlab("Sample") + ylab("Fragments") + ggtitle("Fragments Detected Per Sample")


    # cat("<br>")
    # cat("<br>")
    # cat("How consistently are genes targeted across samples?\n")
    # cat("<br>")
    # barplot(table(rowSums(gcountsf > 0)), main="Consistency of gene coverage", 
    #     ylab='Number of Genes', xlab="Number of samples genes are targeted in")


    cat("<br>")
    cat("<br>")
    cat("Are any genes not covered in most/all samples?\n")
    cat("<br>")
    
    ncidx = order(rowSums(gcounts == 0), decreasing=T)
    pheatmap(log2(gcounts[ncidx,][1:50,] + 1), main="Log2 read coverage, 50 lowest covered",
        cluster_rows=F, scale='none', annotation_col=anncol, fontsize=6, cutree_cols = ncol(gcounts))


    cat("<br>")
    cat("<br>")
    cat("Are any genes not covered in samples from the Final group?\n")
    cat("<br>")

    ncidx = order(rowSums(gcountsF == 0), decreasing=T)
    pheatmap(log2(gcountsF[ncidx,][1:50,] + 1), main="Final: Log2 read coverage, 50 lowest covered",
        cluster_rows=F, scale='none', annotation_col=anncol, fontsize=6, cutree_cols = ncol(gcountsF))


    cat("<br>")
    cat("<br>")
    cat("Final: Are any genes not covered in most/all Final samples after pooling?\n")
    cat("<br>")

    df.combined= data.frame(Final.gDNA = rowSums(gcountsF[, grep("^D", colnames(gcountsF), value=T)]),
                            Final.Plasmid = rowSums(gcountsF[, grep("^F", colnames(gcountsF), value=T)]))

    ncidx = order(rowSums(df.combined == 0), decreasing=T)
    pheatmap(log2(df.combined[ncidx,][1:50,] + 1), main="Final: Log2 read coverage, 50 lowest covered",
        cluster_rows=F, cluster_cols=F, scale='none', fontsize=6, gaps_col = c(1,2))

    cat("<br>")
    cat("<br>")
    cat("Final: Are any genes not covered in most/all Final samples after pooling?\n")
    cat("<br>")
    print(kable(df.combined[ncidx,][1:50,], padding=2) %>%
    kable_styling("striped", full_width = F) %>%
    row_spec(0, angle = 0))

}
```

```{r CountReads, results="hide", echo=FALSE}
# Get sequenced read counts:
readcount = read.table('../read_primer_counts.txt', header=T, as.is=T, sep=' ')
readcount$Sample = gsub('-|_', '.', sapply(strsplit(readcount$sample, '_S'), '[', 1))

# Get sample metadata and setup a standard SampleID
mdata = read.table("../SampleMetaData.csv", header=T, as.is=T, sep='\t')
mdata$Sample = gsub('-|_', '.', mdata$FileID)

# Drop LibC samples from mdata and only quant bams from the A and B Groups
mdata = mdata[mdata$Fragmented.gDNA != 'FC', ]
mdata$Group = factor(c("FA"="Pilot", "FB"="Final")[mdata$Fragmented.gDNA], levels=c("Pilot", "Final"))

bams = paste0("../02-mapped/", mdata$FileID, ".bam")

#bams = dir('../02-mapped/', pattern='*.bam', full.names=T)
gff = '../Reference/AmoebaDB-46_EhistolyticaHM1IMSS.gff'
ann = read.table(gff, header=F, as.is=T, sep='\t', quote='', comment.char='#')
ann = ann[ann$V3 == 'gene',]
ann$gene = gsub('ID=', '', sapply(strsplit(ann$V9, ';'), '[', 1))

#Record ann to contig, start order (sortof chromosome order)
ann = ann[order(ann$V1, ann$V4), ]

forward_stranded_counts = featureCounts(bams, annot.ext = gff, 
    isGTFAnnotationFile=T, 
    GTF.attrType='ID',
    GTF.featureType='gene',
    minOverlap=27,
    allowMultiOverlap=TRUE,
    countMultiMappingReads=TRUE,
    strandSpecific = 1,  #Only reads where R1 is forward W.R.T. the transcript are counted
    isPairedEnd = TRUE,
    nthreads = 7,
    useMetaFeatures = TRUE     #Counts should be summarized by gene
)

gcounts = forward_stranded_counts$counts[, 1:ncol(forward_stranded_counts$counts)]
colnames(gcounts) = gsub('.bam', '', colnames(gcounts))
fscann = cbind(ann[match(rownames(gcounts), ann$gene), ], gcounts)
write.table(fscann, file='forward_stranded_read_counts_by_gene.tsv', sep='\t', row.names=F, col.names=T)


reverse_stranded_counts = featureCounts(bams, annot.ext = gff, 
    isGTFAnnotationFile=T, 
    GTF.attrType='ID',
    GTF.featureType='gene',
    minOverlap=27,
    allowMultiOverlap=TRUE,
    countMultiMappingReads=TRUE,
    strandSpecific = 2,  #Only reads where R1 is reverse W.R.T. the transcript are counted
    isPairedEnd = TRUE,
    nthreads = 7,
    useMetaFeatures = TRUE     #Counts should be summared by gene
)

gcounts = reverse_stranded_counts$counts[, 1:ncol(reverse_stranded_counts$counts)]
colnames(gcounts) = gsub('.bam', '', colnames(gcounts))
rscann = cbind(ann[match(rownames(gcounts), ann$gene), ], gcounts)
write.table(rscann, file='reverse_stranded_read_counts_by_gene.tsv', sep='\t', row.names=F, col.names=T)


unstranded_counts = featureCounts(bams, annot.ext = gff, 
    isGTFAnnotationFile=T, 
    GTF.attrType='ID',
    GTF.featureType='gene',
    minOverlap=27,
    allowMultiOverlap=TRUE,
    countMultiMappingReads=TRUE,
    strandSpecific = 0,     # All reads overlapping 
    isPairedEnd = TRUE,
    nthreads = 7,
    useMetaFeatures = TRUE     #Counts should be summared by gene
)
gcounts = unstranded_counts$counts[, 1:ncol(unstranded_counts$counts)]
colnames(gcounts) = gsub('.bam', '', colnames(gcounts))
uscann = cbind(ann[match(rownames(gcounts), ann$gene), ], gcounts)
write.table(uscann, file='unstranded_read_counts_by_gene.tsv', sep='\t', row.names=F, col.names=T)


```

## Results for forward-read coverage counts:
```{r forward, results = 'asis', echo=FALSE}
plotSummaries(forward_stranded_counts, "forward", mdata)
```

*****

## Results for reverse-read coverage counts:
```{r reverse, results = 'asis', echo=FALSE}
plotSummaries(reverse_stranded_counts, "reverse", mdata)
```

*****

## Results for bi-directional read coverage counts:
```{r unstranded, results = 'asis', echo=FALSE}
plotSummaries(unstranded_counts, "unstranded", mdata)

```



## R session info and packages
```{r sessionInfo}
library("pander")
pander(sessionInfo(), compact = FALSE)
```